### PR TITLE
Adds support for metadata LSP language features in VSCode

### PR DESCRIPTION
**Pull request by dibarbet**
Resolves https://github.com/dotnet/vscode-csharp/issues/5787

In VS, the VS misc workspace delegates to the MetadataAsSource workspace to open the file if it recognizes a metadata file.  This was not happening in VSCode because we do not use the VS misc workspace.  So files would get generated, but never added to the MetadataAsSource workspace.

Instead, I added similar functionality to the LSP misc files workspace (to add/remove from MetadataAsSource if a metadata file is opened/closed).

metadata_lang_features

### CODE REVIEW

File: src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs [hunk 1]
```diff
@@ -292,32 +293,30 @@ public bool ShouldCollapseOnOpen(MetadataAsSourceWorkspace workspace, string fil
         return false;
     }
 
-    public bool TryAddDocumentToWorkspace(MetadataAsSourceWorkspace workspace, string filePath, SourceTextContainer sourceTextContainer)
+    public bool TryAddDocumentToWorkspace(MetadataAsSourceWorkspace workspace, string filePath, SourceTextContainer sourceTextContainer, [NotNullWhen(true)] out DocumentId? documentId)
     {
-        AssertIsMainThread(workspace);
```
Comment (dibarbet) [inline src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs:297]: Wasn't quite sure what to do with all these asserts. The callers of these in the VS layer assert the main thread, but we can't assert the main thread from LSP.  So I removed the asserts.  Not sure if there is a better way.

Comment (jasonmalinowski) [reply src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs:297]: Just put locks around all the methods, if possible? (I haven't looked at the code in detail so this suggestion could be terrible.)

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs:297]: yeah, i agree with jason.  fine to be synchronized.

File: src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs [hunk 2]
```diff
@@ -329,8 +328,7 @@ public bool TryRemoveDocumentFromWorkspace(MetadataAsSourceWorkspace workspace,
 
     private bool RemoveDocumentFromWorkspace(MetadataAsSourceWorkspace workspace, MetadataAsSourceGeneratedFileInfo fileInfo)
     {
-        AssertIsMainThread(workspace);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs:332]: if you also don't like all the synchronous locks, yo ucan make all thsi async with a semaphoreslim, if hte caller can be async+ct.

File: src/Features/Core/Portable/MetadataAsSource/IMetadataAsSourceFileProvider.cs [hunk 3]
```diff
@@ -35,12 +36,16 @@ internal interface IMetadataAsSourceFileProvider
     /// <summary>
     /// Called when the file returned from <see cref="GetGeneratedFileAsync"/> needs to be added to the workspace,
     /// to be opened.  Will be called on the main thread of the workspace host.
+    /// 
+    /// Callers of this must guarantee serial access.
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MetadataAsSource/IMetadataAsSourceFileProvider.cs:40]: don't love this.

File: src/LanguageServer/Protocol/RoslynLanguageServer.cs [hunk 4]
```diff
@@ -99,7 +99,7 @@ protected override IRequestExecutionQueue<RequestContext> ConstructRequestExecut
             // those cases, we do not need to add an additional workspace to manage new files we hear about.  So only
             // add the LspMiscellaneousFilesWorkspace for hosts that have not already brought their own.
             if (serverKind == WellKnownLspServerKinds.CSharpVisualBasicLspServer)
-                AddLazyService<LspMiscellaneousFilesWorkspace>(lspServices => new LspMiscellaneousFilesWorkspace(lspServices, hostServices));
+                AddLazyService<LspMiscellaneousFilesWorkspace>(lspServices => lspServices.GetRequiredService<LspMiscellaneousFilesWorkspaceProvider>().CreateLspMiscellaneousFilesWorkspace(lspServices, hostServices));
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/RoslynLanguageServer.cs:102]: why this pattern?

File: src/LanguageServer/Protocol/Workspaces/LspMiscellaneousFilesWorkspaceProvider.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+/// <summary>
+/// Service to create <see cref="LspWorkspaceManager"/> instances.
+/// This is not exported as a <see cref="ILspServiceFactory"/> as it requires
+/// special base language server dependencies such as the <see cref="HostServices"/>
+/// </summary>
+[ExportCSharpVisualBasicStatelessLspService(typeof(LspMiscellaneousFilesWorkspaceProvider)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class LspMiscellaneousFilesWorkspaceProvider(IMetadataAsSourceFileService metadataAsSourceFileService) : ILspService
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Workspaces/LspMiscellaneousFilesWorkspaceProvider.cs:23]:
```suggestion
internal sealed class LspMiscellaneousFilesWorkspaceProvider(IMetadataAsSourceFileService metadataAsSourceFileService) : ILspService
```

File: src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs [hunk 6]
```diff
@@ -238,7 +238,8 @@ public async Task<(Workspace?, Solution?, TextDocument?)> GetLspDocumentInfoAsyn
                 // As we found the document in a non-misc workspace, also attempt to remove it from the misc workspace
                 // if it happens to be in there as well.
                 if (workspace != _lspMiscellaneousFilesWorkspace)
-                    _lspMiscellaneousFilesWorkspace?.TryRemoveMiscellaneousDocument(uri);
+                    // Do not attempt to remove the file from the metadata workspace (the document is still open).
+                    _lspMiscellaneousFilesWorkspace?.TryRemoveMiscellaneousDocument(uri, removeFromMetadata: false);
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs:242]: surprised that curlies were not required here :)

File: src/LanguageServer/ProtocolUnitTests/Metadata/LspMetadataAsSourceWorkspaceTests.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+using Microsoft.CodeAnalysis.MetadataAsSource;
+using Roslyn.Test.Utilities;
+using Roslyn.Utilities;
+using Xunit;
+using Xunit.Abstractions;
+using LSP = Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.UnitTests.Metadata;
+
+public class LspMetadataAsSourceWorkspaceTests : AbstractLanguageServerProtocolTests
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/ProtocolUnitTests/Metadata/LspMetadataAsSourceWorkspaceTests.cs:18]:
```suggestion
public sealed class LspMetadataAsSourceWorkspaceTests : AbstractLanguageServerProtocolTests
```

**Review by CyrusNajmabadi — APPROVED** [review]
overall ok.  i do think locking of some form would be good.

File: src/LanguageServer/Protocol/RoslynLanguageServer.cs (hunk 4 above)
Comment (dibarbet) [reply src/LanguageServer/Protocol/RoslynLanguageServer.cs:102]: The misc files workspace now requires dependencies from both regular mef (IMetadataAsSourceFileService) and directly from the base language server instance (HostServices) that aren't available as MEF services.

So the provider can import the MEF service normally, then gets passed in the HostServices by the instance.

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by JoeRobich — APPROVED** [review]

**Pull request merged**